File: Sort/quick_sort.py
from typing import List

def quick_sort_range(array: List, first: int, last: int) -> None:
    if first >= last:
        return None

    pivot = array[first]
    pos = last

    for i in range(last, first, -1):
        """ 
        If value greater than pivot,
        swap pos and index to make sure the value of index greater than pos are bigger than pivot 
        """
        if array[i] > pivot: 
            array[i], array[pos] = array[pos], array[i]
            pos -= 1

    # Swap pivot and pos
    array[first], array[pos] = array[pos], array[first]
    # Divided and conquer
    quick_sort_range(array, first, pos-1)  # Left side
    quick_sort_range(array, pos+1, last) # Right side


def quick_sort(array: List) -> None:
    length = len(array)
    first_index = 0
    last_index = length - 1
    quick_sort_range(array, first_index, last_index)


def main() -> None:
    a = [3, 1, 2, 9, 7]
    quick_sort(a)

    print(a)


main()

File: Sort/quick_sort.go
package main

import (
	"fmt"
)

func quickSortHelper(ary []int, left, right int) {
	if left > right {
		return
	}
	pivot := ary[left]
	pos := right
	for i := right; i > left; i-- {
		if ary[i] > pivot {
			ary[i], ary[pos] = ary[pos], ary[i]
			pos--
		}
	}
	ary[left], ary[pos] = ary[pos], ary[left]
	quickSortHelper(ary, left, pos-1)
	quickSortHelper(ary, pos+1, right)
}

func QuickSort(ary []int) {
	if len(ary) == 1 {
		return
	}
	quickSortHelper(ary, 0, len(ary)-1)
}

func main() {
	ary := []int{10, 1, 7}
	fmt.Println("Before sorting", ary)
	QuickSort(ary)
	fmt.Println("After sorting", ary)
}
